File: go/2015-01.go
package main

import (
	"os"
)

func main() {
	runReal()
}

func run(s string) {
	floor := 0
	firstTimeInBasement := 0
	moves := 0
	i := 0
	chars := []rune(s)
	for i = 0; i < len(chars); i++ {
		c := chars[i]
		moves += 1
		if c == '(' {
			floor += 1
		} else {
			floor -= 1
		}
		if firstTimeInBasement == 0 && floor < 0 {
			firstTimeInBasement = moves
		}
	}
	println(floor)
	print(firstTimeInBasement)
}

func runReal() {
	input := readInput()
	run(input)
}

func readInput() string {
	dat, err := os.ReadFile("../input/2015-01.txt")
	check(err)
	return string(dat)
}

func check(e error) {
	if e != nil {
		panic(e)
	}
}

File: ts/2015-01.ts
const fs = require("fs/promises");

const main = async () => {
  const input = await readInput();
  run(input);
};

const run = (s: string) => {
  let floor = 0;
  let firstTimeInBasement = 0;
  let moves = 0;
  const chars = s.split("");

  chars.forEach((c) => {
    moves += 1;
    if (c === "(") {
      floor += 1;
    } else {
      floor -= 1;
      if (firstTimeInBasement == 0 && floor < 0) {
        firstTimeInBasement = moves;
      }
    }
  });

  console.log(floor);
  console.log(firstTimeInBasement);
};

const readInput = async (): Promise<string> => {
  try {
    const data = await fs.readFile("../input/2015-01.txt", {
      encoding: "utf8",
    });
    return data;
  } catch (err) {
    console.log(err);
    return "";
  }
};

main();
